### ISSUE TITLE
nn.Parameter in list

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#189]
I am currently trying to create an (arbitrary) number of parameters in an `nn.Module`. I would store them in a list since `nn.ModuleList` is not intended for `nn.Parameters`, at least thats what the typing suggests. But when using parameters in a list the parent assignment fails. Example:

```python
def test_parameter_usage():
  nn.reset_default_root_name_ctx()
  time = nn.SpatialDim("time")
  feature = nn.FeatureDim("feature", 1)
  data = nn.get_extern_data(nn.Data("data", dim_tags=[nn.batch_dim, time, feature], dtype="float32"))

  class _Net(nn.Module):

    def __init__(self, feat_dim):
      super(_Net, self).__init__()
      self.scales = [nn.Parameter(shape=[feat_dim], trainable=True)]
      self.biases = [nn.Parameter(shape=[feat_dim], trainable=True)]

    def __call__(self, x):
      return self.scales[0] * x + self.biases[0]

  net = _Net(feature)
  x = net(data)
  x.mark_as_default_output()

  config = nn.get_returnn_config().get_complete_py_code_str(net)
```

fails with: 

```
  File "/u/hilmes/experiments/tts_new_sis/recipe/i6_private/users/hilmes/old/tests.py", line 150, in <module>
    line: test_parameter_usage()
    locals:
      test_parameter_usage = <local> <function test_parameter_usage at 0x7ff493430670>
  File "/u/hilmes/experiments/tts_new_sis/recipe/i6_private/users/hilmes/old/tests.py", line 146, in test_parameter_usage
    line: config = nn.get_returnn_config().get_complete_py_code_str(net)
    locals:
      config = <not found>
      nn = <global> <module 'returnn_common.nn' from '/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/__init__.py'>
      nn.get_returnn_config = <global> <function get_returnn_config at 0x7ff452bfef70>
      get_complete_py_code_str = <not found>
      net = <local> <_Net>
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/naming.py", line 763, in ReturnnConfigSerializer.get_complete_py_code_str
    line: return (
            self.get_base_extern_data_py_code_str() +
            self.get_ext_net_dict_py_code_str(
              root_module=root_module, with_imports=False, ref_extern_data_dims_via_global_config=False))
    locals:
      self = <local> <returnn_common.nn.naming.ReturnnConfigSerializer object at 0x7ff452001f70>
      self.get_base_extern_data_py_code_str = <local> <bound method ReturnnConfigSerializer.get_base_extern_data_py_code_str of <returnn_common.nn.naming.ReturnnConfigSerializer object at 0x7ff452001f70>>
      self.get_ext_net_dict_py_code_str = <local> <bound method ReturnnConfigSerializer.get_ext_net_dict_py_code_str of <returnn_common.nn.naming.ReturnnConfigSerializer object at 0x7ff452001f70>>
      root_module = <local> <_Net>
      with_imports = <not found>
      ref_extern_data_dims_via_global_config = <not found>
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/naming.py", line 802, in ReturnnConfigSerializer.get_ext_net_dict_py_code_str
    line: net_dict = self.get_net_dict_raw_dict(root_module=root_module)
    locals:
      net_dict = <not found>
      self = <local> <returnn_common.nn.naming.ReturnnConfigSerializer object at 0x7ff452001f70>
      self.get_net_dict_raw_dict = <local> <bound method ReturnnConfigSerializer.get_net_dict_raw_dict of <returnn_common.nn.naming.ReturnnConfigSerializer object at 0x7ff452001f70>>
      root_module = <local> <_Net>
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/naming.py", line 831, in ReturnnConfigSerializer.get_net_dict_raw_dict
    line: self.name_ctx.prepare_for_config_serialization(root_module=root_module)
    locals:
      self = <local> <returnn_common.nn.naming.ReturnnConfigSerializer object at 0x7ff452001f70>
      self.name_ctx = <local> <NameCtx / module:<_Net>>
      self.name_ctx.prepare_for_config_serialization = <local> <bound method NameCtx.prepare_for_config_serialization of <NameCtx / module:<_Net>>>
      root_module = <local> <_Net>
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/naming.py", line 425, in NameCtx.prepare_for_config_serialization
    line: self._remove_unused_and_assign_parents_and_handle_subnets()
    locals:
      self = <local> <NameCtx / module:<_Net>>
      self._remove_unused_and_assign_parents_and_handle_subnets = <local> <bound method NameCtx._remove_unused_and_assign_parents_and_handle_subnets of <NameCtx / module:<_Net>>>
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/naming.py", line 341, in NameCtx._remove_unused_and_assign_parents_and_handle_subnets
    line: tensor._assign_parent_name_ctx(ref_ctx=root)
    locals:
      tensor = <local> <Parameter 'parameter' [F|F'feature'(1)] via 'variable'>
      tensor._assign_parent_name_ctx = <local> <bound method Tensor._assign_parent_name_ctx of <Parameter 'parameter' [F|F'feature'(1)] via 'variable'>>
      ref_ctx = <not found>
      root = <local> <NameCtx / module:<_Net>>
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/base.py", line 225, in Tensor._assign_parent_name_ctx
    line: assert self.parent_modules  # cannot assign parent without parent modules
    locals:
      self = <local> <Parameter 'parameter' [F|F'feature'(1)] via 'variable'>
      self.parent_modules = <local> []
AssertionError
```

Doing the same without list brackets, (so e.g. `self.scales = nn.Parameter(shape=[feat_dim], trainable=True)` and `return self.scales * x + self.biases`) works without a problem.
I am not sure whether this behavior is somehow intended, because of a different way to store a number of parameters.

**Comment by albertz** [comment]
How does PyTorch handle this? Is this allowed in PyTorch?

**Comment by albertz** [comment]
Btw, what is the use case? If you want to perform some function repeatedly in a loop using the dynamic list of parameters, wouldn't it be more straightforward to make an own module for this function? And then you would use `nn.ModuleList` to store the module list.

**Comment by Atticus1806** [comment]
> How does PyTorch handle this? Is this allowed in PyTorch?

PyTorch has an own `ParameterList`

> Btw, what is the use case?

Right now I am trying to do #188 by hand for now (even though I agree, we should just do a module for it) 

```python
    for (i, module) in enumerate(self.modules):
      x, _ = module(x, in_spatial_dim=time_dim)
      for param in module.parameters():
        param.weight_decay = self.l2

      x = nn.relu(x)
      x = nn.normalize(x, axis=x.feature_dim)
      x = self.scales[i] * x + self.biases[i]
      x = nn.dropout(x, axis=x.feature_dim, dropout=self.dropout[i])
```

This is what I am doing right now. In general this probably is solveable in a Module + Module list (just like the module list I iterate over is), but the module right now is also used without e.g. ReLU so I decided to go like this for now, adding the scale and bias was just a quick first fix.

**Comment by albertz** [comment]
> PyTorch has an own [`ParameterList`](https://pytorch.org/docs/stable/generated/torch.nn.ParameterList.html)

That sounds like a reasonable solution which we could also implement.

Additionally, we should maybe also improve the error message here.

**Comment by albertz** [comment]
So I think we can consider this as solved via #191.
